File: backend/controllers/InvoiceController.js
const Invoice = require("../models/Invoice");
const AppError = require("../utils/AppError");
const Notification = require("../models/Notification");

// Create Invoice
const createInvoice = async (req, res, next) => {
    try {
        const {
            purchase_order_id,
            invoice_number,
            invoice_date,
            status,
        } = req.body;

        const invoice_file = req.file ? req.file.filename : null;

        const result = await Invoice.createInvoice(
            purchase_order_id,
            invoice_number,
            invoice_file,
            invoice_date,
            status
        );
        if (status !== "Paid") {
    const title = "Payment Due";
    const message = `Invoice ${invoice_number} payment is pending.`;

await Notification.createNotificationIfNotExists(
    "Payment Due",
    `Invoice ${invoice_number} payment is pending.`,
    "Payment Due"
);
}

        res.status(201).json({
            success: true,
            message: "Invoice created successfully",
            invoiceId: result.insertId,
        });
    } catch (error) {
        next(error);
    }
};

// Get All Invoices
const getAllInvoices = async (req, res, next) => {
    try {
        const invoices = await Invoice.getAllInvoices();

        res.status(200).json({
            success: true,
            count: invoices.length,
            data: invoices,
        });
    } catch (error) {
        next(error);
    }
};

// Get Invoice By ID
const getInvoiceById = async (req, res, next) => {
    try {
        const invoice = await Invoice.getInvoiceById(req.params.id);

        if (!invoice) {
            return next(new AppError("Invoice not found", 404));
        }

        res.status(200).json({
            success: true,
            data: invoice,
        });
    } catch (error) {
        next(error);
    }
};

// Update Invoice
const updateInvoice = async (req, res, next) => {
    try {
        const {
            purchase_order_id,
            invoice_number,
            invoice_date,
            status,
        } = req.body;

        const existingInvoice = await Invoice.getInvoiceById(req.params.id);

        if (!existingInvoice) {
            return next(new AppError("Invoice not found", 404));
        }

        const invoice_file = req.file
            ? req.file.filename
            : existingInvoice.invoice_file;

        const result = await Invoice.updateInvoice(
            req.params.id,
            purchase_order_id,
            invoice_number,
            invoice_file,
            invoice_date,
            status
        );

        res.status(200).json({
            success: true,
            message: "Invoice updated successfully",
            affectedRows: result.affectedRows,
        });
    } catch (error) {
        next(error);
    }
};

// Mark Invoice as Paid
const markInvoiceAsPaid = async (req, res, next) => {
    try {
        const { id } = req.params;
        const { payment_method, transaction_id } = req.body;

        // Check if invoice exists
        const invoice = await Invoice.getInvoiceById(id);

        if (!invoice) {
            return res.status(404).json({
                success: false,
                message: "Invoice not found",
            });
        }

        await Invoice.markInvoiceAsPaid(
            id,
            "Paid",
            new Date(),
            payment_method,
            transaction_id
        );

        res.status(200).json({
            success: true,
            message: "Invoice marked as paid successfully",
        });
    } catch (error) {
        next(error);
    }
};

// Delete Invoice
const deleteInvoice = async (req, res, next) => {
    try {
        const invoice = await Invoice.getInvoiceById(req.params.id);

        if (!invoice) {
            return next(new AppError("Invoice not found", 404));
        }

        const result = await Invoice.deleteInvoice(req.params.id);

        res.status(200).json({
            success: true,
            message: "Invoice deleted successfully",
            affectedRows: result.affectedRows,
        });
    } catch (error) {
        next(error);
    }
};

module.exports = {
    createInvoice,
    getAllInvoices,
    getInvoiceById,
    updateInvoice,
    markInvoiceAsPaid,
    deleteInvoice,
};

File: backend (1)/backend/src/main/java/com/procureease/backend/controller/InvoiceController.java
package com.procureease.backend.controller;

import com.procureease.backend.dto.CreateInvoiceRequest;
import com.procureease.backend.dto.InvoiceResponse;
import com.procureease.backend.dto.MarkInvoicePaidRequest;
import com.procureease.backend.dto.UpdateInvoiceRequest;
import com.procureease.backend.service.InvoiceService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/invoices")
@RequiredArgsConstructor
public class InvoiceController {

    private final InvoiceService invoiceService;

    // ============================================================
    // Create Invoice
    // ============================================================

    @PostMapping(consumes = "multipart/form-data")
    public ResponseEntity<Map<String, Object>> createInvoice(
            @ModelAttribute CreateInvoiceRequest request,
            @RequestParam("invoice") MultipartFile file
    ) throws IOException {

        InvoiceResponse invoice =
                invoiceService.createInvoice(
                        request,
                        file
                );

        Map<String, Object> response =
                new HashMap<>();

        response.put("success", true);
        response.put(
                "message",
                "Invoice created successfully"
        );
        response.put("data", invoice);

        return ResponseEntity
                .status(201)
                .body(response);
    }

    // ============================================================
    // Get All Invoices
    // ============================================================

    @GetMapping
    public ResponseEntity<Map<String, Object>> getAllInvoices() {

        List<InvoiceResponse> invoices =
                invoiceService.getAllInvoices();

        Map<String, Object> response =
                new HashMap<>();

        response.put("success", true);
        response.put("count", invoices.size());
        response.put("data", invoices);

        return ResponseEntity.ok(response);
    }

    // ============================================================
    // Get Invoice By ID
    // ============================================================

    @GetMapping("/{id}")
    public ResponseEntity<Map<String, Object>> getInvoiceById(
            @PathVariable Integer id
    ) {

        InvoiceResponse invoice =
                invoiceService.getInvoiceById(id);

        Map<String, Object> response =
                new HashMap<>();

        response.put("success", true);
        response.put("data", invoice);

        return ResponseEntity.ok(response);
    }

    // ============================================================
    // Update Invoice
    // ============================================================

    @PutMapping(
            value = "/{id}",
            consumes = "multipart/form-data"
    )
    public ResponseEntity<Map<String, Object>> updateInvoice(
            @PathVariable Integer id,
            @ModelAttribute UpdateInvoiceRequest request,
            @RequestParam(
                    value = "invoice",
                    required = false
            )
            MultipartFile file
    ) throws IOException {

        InvoiceResponse invoice =
                invoiceService.updateInvoice(
                        id,
                        request,
                        file
                );

        Map<String, Object> response =
                new HashMap<>();

        response.put("success", true);
        response.put(
                "message",
                "Invoice updated successfully"
        );
        response.put("data", invoice);

        return ResponseEntity.ok(response);
    }

    // ============================================================
    // Mark Invoice As Paid
    // ============================================================

    @PutMapping("/{id}/pay")
    public ResponseEntity<Map<String, Object>> markInvoiceAsPaid(
            @PathVariable Integer id,
            @RequestBody MarkInvoicePaidRequest request
    ) {

        invoiceService.markInvoiceAsPaid(
                id,
                request
        );

        Map<String, Object> response =
                new HashMap<>();

        response.put("success", true);
        response.put(
                "message",
                "Invoice marked as paid successfully"
        );

        return ResponseEntity.ok(response);
    }

    // ============================================================
    // Delete Invoice
    // ============================================================

    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, Object>> deleteInvoice(
            @PathVariable Integer id
    ) {

        invoiceService.deleteInvoice(id);

        Map<String, Object> response =
                new HashMap<>();

        response.put("success", true);
        response.put(
                "message",
                "Invoice deleted successfully"
        );

        return ResponseEntity.ok(response);
    }
}